heapster-saw/src/Verifier/SAW/Heapster/README.md

```diff
@@ -0,0 +1,183 @@
+
+The goal of the implication prover is to search for a permission implication proof using the implication rules discussed above.
+
+
+Explain algorithm:
+
+- `ImplM` monad captures the form and types of perms on the stack; also has a notion of existential variables that need to be instantiated, some of which already are; display the `ImplM` type vars
+
+- Main entrypoints: `proveVarImpl`, `proveVarsImpl`, and `proveVarsImplAppend`; all written in terms of `proveVarsImplAppend`
+
+- `proveVarsImplAppend`: first tries to do the proof without ending any lifetimes, and if that fails it tries to end lifetimes that could potentially help with the proof
+
+- The "internal-only" functions `proveVarsImplInt` and `proveVarsImplAppendInt`: repeatedly call `findProvablePerm` to choose the "best" permission on the RHS to attempt to prove; briefly explain that will only select a permission whose needed variables are all instantiated
+
+- Next level down is `proveExVarImpl`, which handles the possibility that we are trying to prove a permission `x:p` where `x` itself is an existential variable; this is not really all that common, except for some very specific cases that come up in situtations that are a little too detailed to go into here.
+
+The main logic for proving most of the permission constructs is in the next function down, `proveVarImplInt`, which proves a single permission on a variable `x`. It does this by first pushing the primary permsisions `p` for `x` onto the top of the stack and then proving an implication `x:p |- x:mb_p`, where `mb_p` is the desired permission that we want to prove for `x`. The "mb" prefix indicates that `mb_p` is a permission inside a multi-binding, i.e., a binding of 0 or more existential variables that could be used in the permission we are trying to prove. The `proveVarImplInt` then proceeds by case analysis on `p` and `mb_p`, in most cases using either an introduction rule for proving a construct in `mb_p` on the right or an elimination rule for eliminating a construct in `p` on the left combined with a recursive call to `proveVarImplInt` to prove the remaining implication for smaller `p` and/or smaller `mb_p`. The most complex cases are for proving an equality permission on the right or for proving an implication `x:p1 * ... * pn |- x:p1' * ... * pm'` between conjuncts permissions. For an equality permission on the right, the special-purpose helper function `proveVarEq` is used to prove equality permissions `x:eq(e)` based on the structure of `e`. For proving an implication of conjuncts, the function `proveVarConjImpl` is used, which is structured in a similar manner to `proveVarsImplAppendInt`, in that it repeatedly chooses the "best" permission on the right to prove, proves it, and then recursively proves the remaining permissions on the right until they have all been proved.
+
+The `proveVarAtomicImpl` function is called by `proveVarConjImpl` to prove each conjunct. This function performs 
```

Comment:
This and the next few paragraphs have cut-off sentences. It's fine that this is work in progress, I would just remove the sentence fragments an put TODO's or FIXME's there if you need to remember that spot.

Reply:
Whoops, you're right! Need to fix that...